crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,1528 @@
+            BinaryOps::Sub => self.lhs.derivative(param) - self.rhs.derivative(param),
+            BinaryOps::Mul => self.lhs.derivative(param) * self.rhs.clone() + self.lhs.clone() * self.rhs.derivative(param),
+            BinaryOps::Div => (self.lhs.derivative(param) * self.rhs.clone() - self.lhs.clone() * self.rhs.derivative(param)) / self.rhs.clone() / self.rhs.clone(),
+            BinaryOps::Pow => {
+                if !self.lhs.has_symbol(param.clone()) {
+                    if !self.rhs.has_symbol(param.clone()) {
+                        SymbolExpr::Value( Value::Real(0.0))
+                    } else {
+                        let rhs = SymbolExpr::Unary( Arc::new( Unary{op: UnaryOps::Log, expr: self.lhs.clone()}));
+                        SymbolExpr::Binary( Arc::new( 
+                            Binary{
+                                op: BinaryOps::Mul,
+                                lhs: SymbolExpr::Binary( Arc::new(
+                                    Binary{
+                                        op: BinaryOps::Pow,
+                                        lhs: self.lhs.clone(),
+                                        rhs: self.rhs.clone(),
+                                    } )),
+                                rhs: rhs, }) )
+                    }
+                } else if !self.rhs.has_symbol(param.clone()) {
+                    let rhs = self.rhs.clone() - SymbolExpr::Value( Value::Real(1.0));
+                    self.rhs.clone() * SymbolExpr::Binary( Arc::new( Binary{op: BinaryOps::Pow, lhs: self.lhs.clone(), rhs: rhs}) )
+                } else {
+                    let new_expr = SymbolExpr::Unary( Arc::new( 
+                        Unary { 
+                            op: UnaryOps::Exp,
+                            expr: SymbolExpr::Binary( Arc::new(
+                                Binary{
+                                    op: BinaryOps::Mul,
+                                    lhs: SymbolExpr::Unary( Arc::new( 
+                                        Unary { 
+                                            op: UnaryOps::Log,
+                                            expr: self.lhs.clone(),
+                                        }
+                                    ) ),
+                                    rhs: self.rhs.clone(),
+                                },
+                            ) ),
+                        }
+                    ) );
+                    new_expr.derivative(param)
+                }
+            },
+        }       
+    }
+
+    pub fn eval(&self, recurse: bool) -> Option<Value> {
+        let lval : Value;
+        let rval : Value;
+        if recurse {
+            match self.lhs.eval(recurse) {
+                Some(v) => lval = v,
+                None => return None,
+            }
+            match self.rhs.eval(recurse) {
+                Some(v) => rval = v,
+                None => return None,
+            }

```

Comment:
You could make this a bit less typing by matching the tuple if you would like (no guarantees on the formatting 🙂)
```suggestion
            match (self.lhs.eval(true), self.rhs.eval(true)) {
                (Some(left), Some(right)) => {
                    lval = left;
                    rval = right;
                }
                _ => return None,
            }
```

Reply:
Thank you